zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2019,8 +2020,21 @@ public <T> T retryUntilConnected(final Callable<T> callable)
     }
   }
 
-  private void waitForRetry() {
-    waitUntilConnected(_operationRetryTimeoutInMillis, TimeUnit.MILLISECONDS);
+  /**
+   * ZkClient may have lost the connection to the ZK but can be in cleanup 
+   * stage. Let us make sure that we wait even if the connection appears to 
+   * be connected.
+   */
+  private void waitForRetry(int retryCount) {
+    if (waitUntilConnected(_operationRetryTimeoutInMillis, TimeUnit.MILLISECONDS)) {
+      ExponentialBackoffStrategy retryStrategy =
+        new ExponentialBackoffStrategy(_operationRetryTimeoutInMillis, true);
+      try {
+        Thread.sleep(retryStrategy.getNextWaitInterval(retryCount));
+      } catch (InterruptedException ex) {
+        // we don't need to re-throw.
+      }
+    }
   }
 
   public void setCurrentState(KeeperState currentState) {
```

Comment:
It is odd that `setCurrentState` is a public function. 
1. It is only used in this file itself
2. ZkClient should not allow the connect state be set by others. Let's considering changing it to `private` and visible only to test?

Reply:
I agree that it is public to begin with. 
Can I take that as a follow up or would you prefer to do it as part of this change itself?
thanks

Reply:
Thanks for the reply. Either is ok. :D